device: use iterable sections

gmarull opened the pull request:
Instead of using custom linker code, switch devices to standard iterable sections.

gmarull committed:
toolchain: add STRUCT_SECTION_START|END helpers

Even though they are just wrappers around TYPE_SECTION_START|END, it
allows using the STRUCT_SECTION* API namespace everywhere when working
with struct-based iterable sections.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
device: use iterable sections

Use iterable sections to handle devices list. This simplifies devices
implementation by using standard APIs.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
toolchain: add STRUCT_SECTION_ITERABLE_NAMED

Add a new API that allows to customize the name of an iterable section.
This allows to influence how section elements are sorted.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
shell: device_service: remove levels command

While it may be useful in some contexts, this information can be
obtained at compile time. Removing this command allows to migrate the
device infrastructure code to standard iterable sections, done later.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

nordicjm reviewed (commented):
@fabiobaltieri will this cause any problems for the init checking system you've developed?

gmarull commented:
> @fabiobaltieri will this cause any problems for the init checking system you've developed?

Note that devices still create `struct init_entry` to get initialized. This PR just changes the `struct device` part.

fabiobaltieri commented:
> > @fabiobaltieri will this cause any problems for the init checking system you've developed?
> 
> Note that devices still create `struct init_entry` to get initialized. This PR just changes the `struct device` part.

Yeah it should still work fine as long as that `struct init_entry` is there and the section name follows the expected pattern.

fabiobaltieri commented on include/zephyr/device.h:
```diff
@@ -906,8 +904,9 @@ static inline bool z_impl_device_is_ready(const struct device *dev)
-	const Z_DECL_ALIGN(struct device) DEVICE_NAME_GET(                     \
-		dev_id) Z_DEVICE_SECTION(level, prio) __used =                 \
+	const STRUCT_SECTION_ITERABLE_NAMED(device,                            \
+		Z_DEVICE_SECTION_NAME(level, prio),                            \
```
Does these actually need to be sorted by level and priority?

gmarull replied:
Yes, PM subsystem relies on devices list being sorted. However, what exists in PM subsys for devices, except runtime PM it's pretty useless in many (if not all) applications.

pdgendt replied:
Is the sorting alphabetically? Could this be an issue for single digit priorities?

```
POST_KERNEL1
POST_KERNEL11
POST_KERNEL12
...
POST_KERNEL2
POST_KERNEL21
POST_KERNEL22
...
```

gmarull replied:
@pdgendt you're right, let me try to find a solution.

gmarull replied:
@pdgendt should be fixed now

pdgendt replied:
Nice. Can you add `ITERABLE_SECTION_RAM_NUMERIC` for users that would like the `RAM` variant?

gmarull replied:
@pdgendt actually, current code won't work as levels need to be manually sorted... PRE_KERNEL_1 < APPLICATION. Let me add some iterable helpers that allow to define a section with "subsections".

tejlmand reviewed (dismissed):
temporarily blocking to allow verifying / testing if the proposed changes are compatible with armlink (armclang) implementation.

gmarull committed:
linker: add ITERABLE_SECTION_RA|OM_NUMERIC

In some cases, like devices, it is necessary to sort numerically, that
is, sort 10 before 2. Add a new helper function to do that, up to 2
numeric digits.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

pdgendt commented on include/zephyr/linker/linker-defs.h:
```diff
@@ -48,6 +48,12 @@
 
+#define Z_LINK_ITERABLE_NUMERIC(struct_type) \
+	_CONCAT(_##struct_type, _list_start) = .; \
+	KEEP(*(SORT(._##struct_type.static.*[0-9]*))); \
+	KEEP(*(SORT(._##struct_type.static.*[1-9][0-9]*))); \
```
I don't have a lot of experience with linkers, so ignore if not relevant.

All other defines use `SORT_BY_NAME` explicitly, why not do the same here?
Also, `SORT_BY_INIT_PRIORITY` seems to exist, is this something that could be used instead of a custom numeric implementation?

gmarull replied:
`SORT` is an alias for `SORT_BY_NAME` as per `ld` docs. I just copied other snippets in the same file. I'll try to do some cleanup after (or in parallel) to this PR, so things stay consistent. Regarding `SORT_BY_INIT_PRIORITY`, it says:

> SORT_BY_INIT_PRIORITY is very similar to SORT_BY_NAME. The difference is SORT_BY_INIT_PRIORITY will sort sections into ascending order by numerical value of the GCC init_priority attribute encoded in the section name before placing them in the output file.

ie, `__attribute__ ((init_priority (XXX)))`. I don't think we can use this here.

andyross reviewed (dismissed):
Seems very reasonable, nice boilerplate reduction.

I do still hate the toolchain scheme we have where all the variants (which are, at their core, really very compatible gcc-like things) have to get all the same code, but that's not relevant to this change.

gmarull committed:
init: add INIT_LEVEL_ORD

This patch introduces a new helper: INIT_LEVEL_ORD, which given a level
token, e.g. EARLY, PRE_KERNEL_1, etc. returns its ordinal. This can be
useful for sorting purposes, as level tokens do not result in the right
initialization order if sorted, e.g. POST_KERNEL < APPLICATION, but P >
A.

The implementation uses a bit of macrobatics to achieve this. First, we
define some helpers Z_INIT_XX_XX that evaluate to 1, and where XX
correspond to init levels. With them, we can perform equality checks
like this:

COND_CODE_1(Z_INIT_XX_##level, (VAL_IF_EXISTS), (VAL_IF_NOT_EXISTS))

It's a bit obscure, but it seems to work.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

gmarull committed:
cmake: extensions: add NUMERIC support to zephyr_iterable_section

Add a new flag to zephyr_iterable_section, NUMERIC, that configures
sorting to be numeric (up to 2 digits), that is, 2 comes before e.g. 10.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

pdgendt commented on include/zephyr/linker/linker-defs.h:
```diff
@@ -77,6 +83,21 @@
+ *
+ * @see ITERABLE_SECTION_ROM()
+ */
+#define ITERABLE_SECTION_ROM_NUMERIC(struct_type, subalign) \
```
Add a `RAM` variant while you're at it?

gmarull replied:
done

gmarull commented:
@tejlmand ping, this is blocking future work

galak reviewed (changes requested):
Seeing failures with arm-clang in tests/kernel/device/kernel.device  and other tests.  Will look to debug this to see what's going on.

galak commented on tests/misc/iterable_sections/sections-ram.ld:
```diff
@@ -1,2 +1,3 @@
 ITERABLE_SECTION_RAM(test_ram, 4)
 ITERABLE_SECTION_RAM(test_ram2, 4)
+ITERABLE_SECTION_ROM(test_ram_named, 4)
```
Should probalby be `ITERABLE_SECTION_RAM`

gmarull replied:
fixed

galak reviewed (changes requested):
iterable_section test update is missing updates to tests/misc/iterable_sections/CMakeLists.txt

galak commented on cmake/modules/extensions.cmake:
```diff
@@ -4155,6 +4156,12 @@ function(zephyr_iterable_section)
   endif()
 
+  if(SECTION_NUMERIC)
+    set(INPUT "._${SECTION_NAME}.static.*[0-9];._${SECTION_NAME}.static.*[1-9][0-9]")
```
So for some reason the regex here isn't working as expect for the arm linker.

gmarull replied:
The way it works today:

```
#define CREATE_OBJ_LEVEL(object, level)				\
		__##object##_##level##_start = .;		\
		KEEP(*(SORT(.z_##object##_##level[0-9]_*)));		\
		KEEP(*(SORT(.z_##object##_##level[1-9][0-9]_*)));
```

```
	SECTION_PROLOGUE(initlevel,,)
	{
		/*
		 * link in initialization objects for all objects that are
		 * automatically initialized by the kernel; the objects are
		 * sorted in the order they will be initialized (i.e. ordered
		 * by level, sorted by priority within a level)
		 */
		__init_start = .;
		CREATE_OBJ_LEVEL(init, EARLY)
		CREATE_OBJ_LEVEL(init, PRE_KERNEL_1)
		CREATE_OBJ_LEVEL(init, PRE_KERNEL_2)
		CREATE_OBJ_LEVEL(init, POST_KERNEL)
		CREATE_OBJ_LEVEL(init, APPLICATION)
		CREATE_OBJ_LEVEL(init, SMP)
		__init_end = .;
	} GROUP_ROM_LINK_IN(RAMABLE_REGION, ROMABLE_REGION)
```

vs.

```
zephyr_linker_section(NAME init KVMA RAM_REGION GROUP RODATA_REGION)
zephyr_linker_section_obj_level(SECTION init LEVEL EARLY)
zephyr_linker_section_obj_level(SECTION init LEVEL PRE_KERNEL_1)
zephyr_linker_section_obj_level(SECTION init LEVEL PRE_KERNEL_2)
zephyr_linker_section_obj_level(SECTION init LEVEL POST_KERNEL)
zephyr_linker_section_obj_level(SECTION init LEVEL APPLICATION)
zephyr_linker_section_obj_level(SECTION init LEVEL SMP)
```

```cmake
function(zephyr_linker_section_obj_level)
...
  zephyr_linker_section_configure(
    SECTION ${OBJ_SECTION}
    INPUT ".z_${OBJ_SECTION}_${OBJ_LEVEL}?_"
    SYMBOLS __${OBJ_SECTION}_${OBJ_LEVEL}_start
    KEEP SORT NAME
  )
  zephyr_linker_section_configure(
    SECTION ${OBJ_SECTION}
    INPUT ".z_${OBJ_SECTION}_${OBJ_LEVEL}??_"
    KEEP SORT NAME
  )
```

So it looks like the cmake generator doesn't share the same syntax... Is there any useful documentation I can refer to? We need to clean all this cmake/ld duplication mess, changing something there is like gambling now...

gmarull replied:
As far as I can see, armlink doesn't support the `[...]` character range syntax: https://developer.arm.com/documentation/dui0804/a/CJAJJGEH. First let me try to align cmake/ld scripts and find an alternative solution.

gmarull commented:
> iterable_section test update is missing updates to tests/misc/iterable_sections/CMakeLists.txt

done

gmarull committed:
toolchain: add trailing _ to iterable section name

Add a trailing _ to the section iterable name. This is useful, for
example, to implement numeric sorting like this:

SORT(.z_device_LEVEL_?_)
SORT(.z_device_LEVEL_??_)

Without the trailing _ it would not be possible to use the ?? wildcard
without triggering into trigraphs issues, because linker-defs.h header
is included in C files as well.

Signed-off-by: Gerard Marull-Paretas <gerard.marull@nordicsemi.no>

galak commented on tests/misc/iterable_sections/src/main.c:
```diff
@@ -98,6 +119,12 @@ const STRUCT_SECTION_ITERABLE_NAMED(test_rom_named, C, rom7) = {0x30};
 
+/* declare in random order to check that the linker is sorting numerically */
+const STRUCT_SECTION_ITERABLE(test_rom_numeric, rom10) = {0x10};
+const STRUCT_SECTION_ITERABLE(test_rom_numeric, rom12) = {0x30};
```
Can  you add a number that is 0-9 in this test to make sure its sorted correctly.

galak reviewed (changes requested):
Builds & runs on arm-clang now.  One comment on improving the testcase.  Otherwise LGTM.

gmarull commented:
Note: compliance check is a false positive (? formatting in linker-defs.h)

evgeniy-paltsev reviewed (changes requested):
Build & run tests with ARC MWDT.

gmarull commented:
> ARC MWDT

Am I supposed to order the toolchain?

![image](https://github.com/zephyrproject-rtos/zephyr/assets/25011557/2123b078-7b80-432c-b1e7-f5c37db536db)

I think that proprietary toolchains should be part of the official CI, otherwise, PRs like this can take unnecessary long time to get merged.

nashif commented:
> Build & run tests with ARC MWDT.

at least provide some context and more details about what is failing and how it can be fixed. You obviously know that you are a mong a few who can build with this toolchain.

gmarull commented:
> I've tested this with ARC MWDT toolchain - each configutartion now have ~15 tests failing in runtime.
> 
> I.e, on `nsim_hs` (simulation-based platform) we have following tests failing:
> 
> ```
> samples/userspace/prod_consumer/sample.userspace.prod_consumer
> samples/application_development/out_of_tree_driver/sample.drivers.out_of_tree
> samples/drivers/crypto/sample.drivers.crypto.mbedtls.micro
> samples/drivers/crypto/sample.drivers.crypto.mbedtls
> tests/kernel/device/kernel.device.pm
> tests/misc/iterable_sections/misc.iterable_sections
> tests/kernel/device/kernel.device
> tests/lib/devicetree/devices/libraries.devicetree.devices
> tests/drivers/eeprom/shell/drivers.eeprom.shell
> tests/drivers/ipm/drivers.ipc.mailbox
> tests/drivers/can/shell/drivers.can.shell
> tests/drivers/sensor/generic/drivers.sensor.generic.fpu
> tests/drivers/sensor/generic/drivers.sensor.generic
> tests/subsys/pm/device_runtime_api/pm.device_runtime.api
> ```
> 
> I need some time to investigate this

can these failures be bisected to any of the commits in this PR? Do you have any logs so I can also take a look?

gmarull commented:
> Ok, I've done bisecting and the isse appears on the last commit ([teslabs@c550760](https://github.com/teslabs/zephyr/commit/c55076023796fa0fc0313e514c94f3ddca61f55f)) where we actually enable iterable sections - so it isn't really helpful.

Do iterable sections work when using this toolchain?

evgeniy-paltsev commented:
In all the cases the symphtoms looks similar - we try to access missing (unitialized?) device.

I.e:

```
/path/twister-out/nsim_hs/samples/drivers/crypto/sample.drivers.crypto.mbedtls/handler.log
--------------------
[0/1] nSIM COMMAND: /path/nsimdrv -propsfile /path/zephyr/boards/arc/nsim/support/nsim_hs.props /path/twister-out/nsim_hs/samples/drivers/crypto/sample.drivers.crypto.mbedtls/zephyr/zephyr.elf
Console now belongs to UART, hit CRTL-] to return to simulator.
*** Booting Zephyr OS build zephyr-v2.4.0-34232-gc55076023796 ***
E: CRYPTO_MTLS pseudo device not found
```

Other tests logs:
[err-iter-sections.log](https://github.com/zephyrproject-rtos/zephyr/files/11443320/err-iter-sections.log)

gmarull commented:
> In all the cases the symphtoms looks similar - we try to access missing (unitialized?) device.
> 
> I.e:
> 
> ```
> /path/twister-out/nsim_hs/samples/drivers/crypto/sample.drivers.crypto.mbedtls/handler.log
> --------------------
> [0/1] nSIM COMMAND: /path/nsimdrv -propsfile /path/zephyr/boards/arc/nsim/support/nsim_hs.props /path/twister-out/nsim_hs/samples/drivers/crypto/sample.drivers.crypto.mbedtls/zephyr/zephyr.elf
> Console now belongs to UART, hit CRTL-] to return to simulator.
> *** Booting Zephyr OS build zephyr-v2.4.0-34232-gc55076023796 ***
> E: CRYPTO_MTLS pseudo device not found
> ```
> 
> Other tests logs: [err-iter-sections.log](https://github.com/zephyrproject-rtos/zephyr/files/11443320/err-iter-sections.log)

Can you run the iterable_sections test alone, from `main`?

evgeniy-paltsev commented:
linker_zephyr_pre0.cmd and some other artefacts:

[misc.iterable_sections.zip](https://github.com/zephyrproject-rtos/zephyr/files/11443798/misc.iterable_sections.zip)

evgeniy-paltsev commented:
@gmarull, ok I see one potential issue (for MWDT only), it's same as we have here: #57756 but for Z_LINK_ITERABLE_NUMERIC

gmarull commented:
@evgeniy-paltsev please check latest revision (probably apply https://github.com/zephyrproject-rtos/zephyr/pull/57756 as well)

evgeniy-paltsev commented:
> @evgeniy-paltsev please check latest revision (probably apply https://github.com/zephyrproject-rtos/zephyr/pull/57756 as well)

Testing it RN

evgeniy-paltsev commented:
Hi @gmarull,

I'm wondering if the change you've done in https://github.com/zephyrproject-rtos/zephyr/pull/57597 is required for this PR?

I've noticed that it actully breaks platforms with ARC GNU toolchain (haven't previoully noted it as was looking on the 100% of failing tests with ARC MWDT :) - so looks like it changes initialization order.

evgeniy-paltsev reviewed (approved):
Ok, now with the changes from #57756 are cherry-picked to this everything seems to working fine.

The issues with ARC GNU toolchain have different rootcause and not related to this PR.

carlescufi commented:
CI is a false positive.
